lib/AST/Macro.cpp

```diff
@@ -880,6 +880,108 @@ MacroRange MacroSubstitution::ReplacementChildren(void) const noexcept {
   }
 }
 
+// Returns the first and last fully substituted tokens from a given macro
+// substitution, if they exist.
+std::pair<std::optional<Token>, std::optional<Token>>
+GetFirstAndLastFullySubstitutedTokens(
+  MacroSubstitution const &substitution) noexcept {
+
+  // Get the fully substituted first token
+  std::optional<Token> first_tok = std::nullopt;
+  {
+    auto subst = std::optional(substitution);
+    auto first_replacement_child = std::optional<Macro>();
+    while (subst && !subst->ReplacementChildren().empty()) {
+      // Traverse the next begin token
+      first_replacement_child = *subst->ReplacementChildren().begin();
+      subst = MacroSubstitution::From(*first_replacement_child);
+    }
+    if (first_replacement_child) {
+      const auto first_macro_tok = MacroToken::From(*first_replacement_child);
+      if (first_macro_tok) {
+        first_tok = std::optional(first_macro_tok->ParsedLocation());
+      }
+    }
+  }
+
+  // Get the last substituted first token
+  std::optional<Token> last_tok = std::nullopt;
+  {
+    auto subst = std::optional(substitution);
+    auto last_replacement_child = std::optional<Macro>();
+    while (subst && !subst->ReplacementChildren().empty()) {
+      // Traverse the next end token
+      last_replacement_child = *(subst->ReplacementChildren().end() - 1);
+      subst = MacroSubstitution::From(*last_replacement_child);
+    }
+    if (last_replacement_child) {
+      const auto last_macro_tok = MacroToken::From(*last_replacement_child);
+      if (last_macro_tok) {
+        last_tok = std::optional(last_macro_tok->ParsedLocation());
```

Comment:
Don't need `std::optional` here.

Reply:
Why not? If a macro expands to no tokens then is it possible for the the first/last token to be nothing?